### ISSUE TITLE
Zipformer LoRA - nan gradients

### ISSUE THREAD

**Issue by divyeshrajpura4114** [k2-fsa/icefall#2024]
I have trained model using zipformer_lora receipe and now I am fine-tuning it to adapt for new use-case. I am seeing lot of grads are nan (however, the training is running without any error)

Is this expected ?

```
2025-09-16 12:48:58,748 WARNING [optim.py:726] Parameter dominating tot_sumsq encoder.encoders.5.encoder.layers.1.feed_forward3.in_proj.lora_A with proportion 0.11, where dominant_sumsq=(grad_sumsq*orig_rms_sq)=7.541e-05, grad_sumsq=1.008e-02, orig_rms_sq=7.478e-03                                                                                                       
2025-09-16 12:48:58,760 WARNING [optim.py:665] Parameters with most larger-than-usual grads, with ratios, are: [(nan, 'encoder_embed.out_norm.log_scale', []), (nan, 'encoder.encoders.0.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.0.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.1.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.1.enco
der.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.2.norm.log_scale', []), (nan, 'encoder.encoders.3.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.3.encoder.layers.1.norm.log_scale', [])]
2025-09-16 12:48:59,738 WARNING [optim.py:604] Scaling gradients by 0.13609939813613892, model_norm_threshold=0.0031250857282429934                                                                                                                                                                               
2025-09-16 12:48:59,813 WARNING [optim.py:726] Parameter dominating tot_sumsq encoder.encoders.1.encoder.layers.1.feed_forward1.out_proj.l.lora_B with proportion 0.03, where dominant_sumsq=(grad_sumsq*orig_rms_sq)=1.594e-05, grad_sumsq=1.323e+05, orig_rms_sq=1.205e-10                                                                                                    
2025-09-16 12:48:59,824 WARNING [optim.py:665] Parameters with most larger-than-usual grads, with ratios, are: [(nan, 'encoder_embed.out_norm.log_scale', []), (nan, 'encoder.encoders.0.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.0.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.1.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.1.enco
der.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.2.norm.log_scale', []), (nan, 'encoder.encoders.3.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.3.encoder.layers.1.norm.log_scale', [])]
2025-09-16 12:49:00,846 WARNING [optim.py:588] Clipping_scale=2.0, grad-norm quartiles 5.651e-03 1.010e-02 1.581e-02 2.146e-02 3.004e-02, threshold=3.162e-02, percent-clipped=100.0                                                                                                                              
2025-09-16 12:49:04,898 INFO [scaling.py:1130] Whitening: name=encoder.encoders.2.encoder.layers.1.feed_forward2.out_whiten, num_groups=1, num_channels=384, metric=6.13 vs. limit=15.0                                                                                                                           
2025-09-16 12:49:05,372 INFO [scaling.py:219] ScheduledFloat: name=encoder.encoders.2.encoder.layers.0.self_attn_weights.pos_emb_skip_rate, batch_count=62836.666666666664, ans=0.0                                                   
2025-09-16 12:49:07,338 INFO [scaling.py:219] ScheduledFloat: name=encoder.encoders.0.layers.1.feed_forward1.out_proj.dropout.p, batch_count=62836.666666666664, ans=0.1                                                                                              
2025-09-16 12:49:19,755 INFO [train_finetune.py:1278] Epoch 3, batch 50, loss[loss=0.942, simple_loss=0.819, pruned_loss=0.404, ctc_loss=0.6428, over 22263.00 frames. ], tot_loss[loss=0.8907, simple_loss=0.7771, pruned_loss=0.3801, ctc_loss=0.6103, over 972079.28 frames. ], batch size: 226, lr: 6.83e-03, grad_scale: 8.0
2025-09-16 12:49:39,828 INFO [scaling.py:219] ScheduledFloat: name=encoder.encoders.3.encoder.layers.0.feed_forward3.hidden_balancer.prob, batch_count=62903.333333333336, ans=0.125                                                                                                                              
2025-09-16 12:49:40,487 INFO [scaling.py:1130] Whitening: name=encoder.encoders.0.layers.0.feed_forward2.out_whiten, num_groups=1, num_channels=192, metric=6.35 vs. limit=15.0                                                                                                                                   
2025-09-16 12:49:42,742 WARNING [optim.py:604] Scaling gradients by 0.4466124176979065, model_norm_threshold=0.031620532274246216                                                                                                                                     
2025-09-16 12:49:42,819 WARNING [optim.py:726] Parameter dominating tot_sumsq encoder.encoders.5.encoder.layers.1.feed_forward3.in_proj.lora_A with proportion 0.07, where dominant_sumsq=(grad_sumsq*orig_rms_sq)=3.345e-04, grad_sumsq=2.288e-02, orig_rms_sq=1.462e-02                                                                                                       
2025-09-16 12:49:42,830 WARNING [optim.py:665] Parameters with most larger-than-usual grads, with ratios, are: [(nan, 'encoder_embed.out_norm.log_scale', []), (nan, 'encoder.encoders.0.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.0.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.1.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.1.enco
der.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.2.norm.log_scale', []), (nan, 'encoder.encoders.3.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.3.encoder.layers.1.norm.log_scale', [])]
2025-09-16 12:49:45,906 WARNING [optim.py:604] Scaling gradients by 0.48775166273117065, model_norm_threshold=0.031620532274246216                                                                                                                                                                                
2025-09-16 12:49:45,985 WARNING [optim.py:726] Parameter dominating tot_sumsq encoder.encoders.5.encoder.layers.1.feed_forward3.in_proj.lora_A with proportion 0.06, where dominant_sumsq=(grad_sumsq*orig_rms_sq)=2.380e-04, grad_sumsq=1.628e-02, orig_rms_sq=1.462e-02                                                                                                       
2025-09-16 12:49:45,996 WARNING [optim.py:665] Parameters with most larger-than-usual grads, with ratios, are: [(nan, 'encoder_embed.out_norm.log_scale', []), (nan, 'encoder.encoders.0.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.0.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.1.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.1.enco
der.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.2.norm.log_scale', []), (nan, 'encoder.encoders.3.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.3.encoder.layers.1.norm.log_scale', [])]
2025-09-16 12:49:47,039 WARNING [optim.py:604] Scaling gradients by 0.452959805727005, model_norm_threshold=0.031620532274246216                                                                                                                                                                                  
2025-09-16 12:49:47,118 WARNING [optim.py:726] Parameter dominating tot_sumsq encoder.encoders.5.encoder.layers.1.feed_forward3.in_proj.lora_A with proportion 0.08, where dominant_sumsq=(grad_sumsq*orig_rms_sq)=3.661e-04, grad_sumsq=2.375e-02, orig_rms_sq=1.541e-02                                         
2025-09-16 12:49:47,129 WARNING [optim.py:665] Parameters with most larger-than-usual grads, with ratios, are: [(nan, 'encoder_embed.out_norm.log_scale', []), (nan, 'encoder.encoders.0.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.0.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.1.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.1.enco
der.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.2.norm.log_scale', []), (nan, 'encoder.encoders.3.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.3.encoder.layers.1.norm.log_scale', [])]
2025-09-16 12:49:48,326 WARNING [optim.py:604] Scaling gradients by 0.3981519937515259, model_norm_threshold=0.031620532274246216                                                                                                                                                                                 
2025-09-16 12:49:48,404 WARNING [optim.py:726] Parameter dominating tot_sumsq encoder.encoders.5.encoder.layers.1.feed_forward3.in_proj.lora_A with proportion 0.06, where dominant_sumsq=(grad_sumsq*orig_rms_sq)=3.883e-04, grad_sumsq=2.519e-02, orig_rms_sq=1.541e-02                                         
2025-09-16 12:49:48,415 WARNING [optim.py:665] Parameters with most larger-than-usual grads, with ratios, are: [(nan, 'encoder_embed.out_norm.log_scale', []), (nan, 'encoder.encoders.0.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.0.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.1.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.1.enco
der.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.1.norm.log_scale', []), (nan, 'encoder.encoders.2.encoder.layers.2.norm.log_scale', []), (nan, 'encoder.encoders.3.encoder.layers.0.norm.log_scale', []), (nan, 'encoder.encoders.3.encoder.layers.1.norm.log_scale', [])]
2025-09-16 12:49:49,434 WARNING [optim.py:604] Scaling gradients by 0.4268186092376709, model_norm_threshold=0.031620532274246216                                                                                                                                                                                 
2025-09-16 12:49:49,514 WARNING [optim.py:726] Parameter dominating tot_sumsq encoder.encoders.5.encoder.layers.1.feed_forward3.in_proj.lora_A with proportion 0.08, where dominant_sumsq=(grad_sumsq*orig_rms_sq)=4.268e-04, grad_sumsq=2.769e-02, orig_rms_sq=1.541e-02
```

Here is my environment configuration
```
  "env_info": {
    "icefall-git-branch": "main",
    "icefall-git-date": "Tue Sep 16 11:09:49 2025",
    "icefall-git-sha1": "fbd1854-dirty",
    "icefall-path": "/workspace/icefall",
    "k2-build-type": "Release",
    "k2-git-date": "Mon Jul 14 07:51:57 2025",
    "k2-git-sha1": "9399d1b01a6309e54b62d885e93209bcd66c1e7d",
    "k2-path": "/opt/conda/lib/python3.11/site-packages/k2/__init__.py",
    "k2-version": "1.24.4",
    "k2-with-cuda": true,
    "lhotse-path": "/opt/conda/lib/python3.11/site-packages/lhotse/__init__.py",
    "lhotse-version": "1.31.0.dev+git.273e312.clean",
    "python-version": "3.11",
    "torch-cuda-available": true,
    "torch-cuda-version": "12.4",
    "torch-version": "2.6.0+cu124"
  },
```

**Comment by divyeshrajpura4114** [comment]
Thanks @marcoyang1998 

We have done objective evaluation on internal test set, and no improvement are observed after fine-tuning. 

We are moving ahead with other approaches, closing this.
